app/lib/crosstables_fetcher.rb

```diff
@@ -2,12 +2,63 @@ module CrosstablesFetcher
 
   BASE_URL = "https://www.cross-tables.com/"
 
+  # Hold event data scraped from the main page
+  Event = Struct.new(:xtables_id, :location, :date)
+
   module_function
 
-  def get_players(xtables_id)
-    doc = fetch_url("entrants.php?u=#{xtables_id}")
+  # ---------------------------------------
+  # Main entry points to download and parse data
+
+  def get_all_naspa_events
+    doc = fetch_main_page()
+    events = extract_naspa_events(doc)
+    events.map do |ev|
+      ev_doc = fetch_entrants_page(ev.xtables_id)
+      event = extract_event(ev_doc)
+      event.update(event.to_h)
+    end
+  end
+
+  def get_players(event_id)
+    doc = fetch_entrants_page(event_id)
+    extract_players(doc)
+  end
+
+  # Grabs the player's pic based on the player's xtables ID
+  def get_player_photo(player_id)
+    doc = fetch_player_page(player_id)
+    photo_path = extract_photo_path(doc)
+    URI.open(photo_path)
+  end
+
+  # ---------------------------------------
+  # Fetch pages from cross-tables.com
+
+  def fetch_url(relative_url)
+    url = "#{BASE_URL}/#{relative_url}"
+    html = URI.open(url)
+    doc = Nokogiri::HTML(html)
+  end
+
+  def fetch_main_page
+    fetch_url("")
+  end
+
+  def fetch_entrants_page(event_id)
+    fetch_url("/entrants.php?u=#{event_id}")
```

Comment:
does this work? I think it should be no slash, given the base url

Reply:
good catch! fixing

Reply:
done